generators/chipyard/src/main/scala/harness/HasHarnessInstantiators.scala

```diff
@@ -0,0 +1,96 @@
+package chipyard.harness
+
+import chisel3._
+
+import scala.collection.mutable.{ArrayBuffer, LinkedHashMap}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import org.chipsalliance.cde.config.{Field, Parameters, Config}
+import freechips.rocketchip.util.{ResetCatchAndSync}
+import freechips.rocketchip.prci.{ClockBundle, ClockBundleParameters, ClockSinkParameters, ClockParameters}
+import freechips.rocketchip.stage.phases.TargetDirKey
+
+import chipyard.harness.{ApplyHarnessBinders, HarnessBinders}
+import chipyard.iobinders.HasIOBinders
+import chipyard.clocking.{SimplePllConfiguration, ClockDividerN}
+import chipyard.{ChipTop}
+
+// -------------------------------
+// Chipyard Test Harness
+// -------------------------------
+
+case object MultiChipNChips extends Field[Int](0) // 0 means ignore MultiChipParams
+case class MultiChipParameters(chipId: Int) extends Field[Parameters]
+case object BuildTop extends Field[Parameters => LazyModule]((p: Parameters) => new ChipTop()(p))
+case object HarnessClockInstantiatorKey extends Field[() => HarnessClockInstantiator]()
+case object HarnessBinderClockFrequencyKey extends Field[Double](100.0) // MHz
+case object MultiChipIdx extends Field[Int](0)
+
+class WithMultiChip(id: Int, p: Parameters) extends Config((site, here, up) => {
+  case MultiChipParameters(`id`) => p
```

Comment:
Can you give me a brief primer on how CDE keys can have parameters? Maybe I'm forgetting this exists but this seems new?

Reply:
CDE have always supported parameters, but they are very rarely used.
To make this work, you have to define the key as a `case class SomeKey(someParams: ParamsType) extends Field[...]` instead of `case object SomeKey extends Field[...]`.

The only other trick is when matching on it, if you need to match on a variable defined outside the case expression, you have to put the variable in backticks... ``case SomeKey(`params`)`` instead of `case SomeKey(params)`